File: next/dweb-browser/Helper/src/State.cs
namespace DwebBrowser.Helper;


static class StateShared
{
    internal static List<StateBase> ObsStack = new();
}

public abstract class StateBase
{
    /// <summary>
    /// 我的依赖：当依赖更新，我需要重新执行更新
    /// </summary>
    protected HashSet<StateBase> _Deps = new();
    public IReadOnlySet<StateBase> Deps { get => _Deps; }
    /// <summary>
    /// 我的引用：当我更新，我需要去通知它们执行更新
    /// </summary>
    protected HashSet<StateBase> _Refs = new();
    public IReadOnlySet<StateBase> Refs { get => _Refs; }

    public bool AddDep(StateBase dep)
    {
        var success = _Deps.Add(dep);
        if (success)
        {
            dep.AddRef(this);
        }
        return success;
    }
    public bool AddRef(StateBase @ref)
    {
        var success = _Refs.Add(@ref);
        if (success)
        {
            @ref.AddDep(this);
        }
        return success;
    }
    public bool RmDep(StateBase dep)
    {
        var success = _Deps.Remove(dep);
        if (success)
        {
            dep.RmRef(this);
        }
        return success;
    }
    public bool RmRef(StateBase @ref)
    {
        var success = _Refs.Remove(@ref);
        if (success)
        {
            @ref.RmDep(this);
        }
        return success;
    }
    public void ClearDeps()
    {
        foreach (var dep in _Deps.ToArray())
        {
            RmDep(dep);
        }
    }

    public abstract void BeGet(bool? force = null);
}
public class State<T> : StateBase
{


    Func<T> getter;
    Func<T, bool> setter;
    T? cache;
    bool hasCache = false;
    public State(Func<T> getter, Func<T, bool> setter)
    {
        this.getter = getter;
        this.setter = setter;
    }

    public State(Func<T> getter) : this(getter, (_) => false)
    {
    }

    public State(T defaultValue) : this(() => defaultValue, (newValue) =>
    {
        if (defaultValue!.Equals(newValue))
        {
            return false;
        }
        defaultValue = newValue;
        return true;
    })
    {
        cache = defaultValue;
        hasCache = true;
    }

    public event Signal<T, T?>? OnChange;

    public Task<T> GetNext()
    {
        var res = new PromiseOut<T>();
        this.OnChange += async (data, old, self) =>
        {
            res.Resolve(data);
            this.OnChange -= self;
        };
        return res.WaitPromiseAsync();
    }

    public async Task Until(Func<T, bool> checker)
    {
        if (checker(Get()))
        {
            return;
        }
        var locker = new PromiseOut<T>();
        this.OnChange += async (data, old, self) =>
        {
            if (checker(data)) {
                this.OnChange -= self;
                locker.Resolve(data);
            }
        };
        await locker.WaitPromiseAsync();
    }

    public T Get(bool? force = null)
    {
        lock (StateShared.ObsStack)
        {
            if (StateShared.ObsStack.LastOrDefault() is not null and var caller)
            {
                /// 将调用者存储到自己的依赖中
                caller.AddDep(this);
            }
        }
        if (force is null)
        {
            force = !hasCache;
        }


        if (force is true)
        {
            /// 自己也将作为调用者
            /// 
            lock (StateShared.ObsStack)
            {
                StateShared.ObsStack.Add(this);
            }
            /// 调用之前，清空自己的依赖，重新收集依赖
            ClearDeps();
            try
            {
                var oldValue = cache;
                cache = getter();
                hasCache = true;
                _ = OnChange?.Emit(cache, oldValue);
                return cache;
            }
            finally
            {
                /// 移除自己作为调用者的身份
                lock (StateShared.ObsStack)
                {
                    StateShared.ObsStack.Remove(this);
                }
            }
        }
        return cache!;
    }

    public bool Set(T value, bool force = false)
    {
        if (setter(value) || force)
        {
            /// 强制更新值
            Get(true);
            /// 向自己的调用者发去通知
            foreach (var @ref in _Refs.ToArray())
            {
                @ref.BeGet(true);
            }
            return true;
        }
        return false;
    }

    public bool Update(Func<T?, T> updater, bool force = true)
    {
        return Set(updater(cache), force);
    }
    public bool Update(Func<T?, bool> updater)
    {
        return cache is not null ? Set(cache, updater(cache)) : false;
    }
    public bool Update(Action<T?> updater, bool force = true)
    {
        updater(cache);
        return cache is not null ? Set(cache, force) : false;
    }

    public async IAsyncEnumerable<T> ToStream()
    {
        PromiseOut<T>? waitter = null;
        Mutex locker = new();

        Queue<T> cacheList = new();
        Signal<T, T?> onChange = async (value, old, _) =>
        {

            locker.WaitOne();

            try
            {
                if (waitter is not null)
                {
                    waitter.Resolve(value);
                    waitter = null;
                    return;
                }

                cacheList.Append(value);
            }
            finally
            {
                locker.ReleaseMutex();
            }
        };
        try
        {
            OnChange += onChange;

            yield return Get();
            while (true)
            {

                while (true)
                {
                    bool success;
                    T? item;
                    locker.WaitOne();
                    try
                    {
                        success = cacheList.TryDequeue(out item);
                    }
                    finally
                    {
                        locker.ReleaseMutex();
                    }
                    if (!success)
                    {
                        break;
                    }
                    yield return item!;
                }

                locker.WaitOne();
                PromiseOut<T> w;
                waitter = w = new();
                locker.ReleaseMutex();
                yield return await w.WaitPromiseAsync();
            }
        }
        finally
        {
            OnChange -= onChange;
            locker.Dispose();
        }
    }

    public override void BeGet(bool? force = null)
    {
        Get(force);
    }
}

File: android/helper/src/main/java/org/dweb_browser/helper/State.kt
package org.dweb_browser.helper

import kotlinx.coroutines.InternalCoroutinesApi
import kotlinx.coroutines.internal.synchronized
import kotlinx.coroutines.runBlocking
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import java.util.Collections


object StateShared {
  val obsStack = mutableListOf<StateBase>()
}

abstract class StateBase {

  val deps = Collections.synchronizedSet(HashSet<StateBase>());

  val refs = Collections.synchronizedSet(HashSet<StateBase>());

  fun addDep(dep: StateBase): Boolean {
    val success = deps.add(dep)
    if (success) {
      dep.addRef(this)
    }
    return success
  }

  fun addRef(ref: StateBase): Boolean {
    val success = refs.add(ref)
    if (success) {
      ref.addDep(this)
    }
    return success
  }

  fun rmDep(dep: StateBase): Boolean {
    val success = deps.remove(dep)
    if (success) {
      dep.rmRef(this)
    }
    return success
  }

  fun rmRef(ref: StateBase): Boolean {
    val success = refs.remove(ref)
    if (success) {
      ref.rmDep(this)
    }
    return success
  }

  fun clearDeps() {
    deps.toTypedArray().forEach { dep ->
      rmDep(dep)
    }
  }

  abstract suspend fun beGet(force: Boolean? = null)
}

@Suppress("UNCHECKED_CAST")
class State<T>(private var defaultValue: T) : StateBase() {

  fun setter(newValue: T): Boolean {
    cache = defaultValue;
    hasCache = true;

    return if (this.defaultValue == newValue) false
    else {
      defaultValue = newValue
      true
    }
  }

  fun getter(): T {
    // 如果是闭包
    if (defaultValue is Function<*>) {
      val runFun = defaultValue as () -> T
      return runFun()
    }
    return defaultValue
  }

  init {
    setter(defaultValue)
  }

  var cache: T? = null
  private var hasCache = false

  // 第一个t是新值，第二个t是旧值
  private val changeSignals = Signal<Pair<T, T?>>()

  var onChange = changeSignals.toListener()

  suspend fun getNext(): T {
    val res = PromiseOut<T>()
    onChange { newData ->
      res.resolve(newData.first)
    }
    return res.waitPromise()
  }

  suspend fun until(checker: (T) -> Boolean): T? {
    if (checker(get())) {
      return null
    }
    val locker = PromiseOut<T>()
    onChange { newData ->
      locker.resolve(newData.first)
    }
    return locker.waitPromise()
  }

  @OptIn(InternalCoroutinesApi::class)
  fun get(force: Boolean? = null): T {
    synchronized(StateShared.obsStack) {
      val caller = StateShared.obsStack.lastOrNull()
      if (caller !== null) {
        caller.addDep(this)
      }
    }
    val inForce = if (force == null) hasCache else true
    if (inForce) {
      /// 自己也将作为调用者
      synchronized(StateShared.obsStack) {
        StateShared.obsStack.add(this)
      }
      /// 调用之前，清空自己的依赖，重新收集依赖
      clearDeps()
      try {
        val oldValue = cache
        cache = getter()
        hasCache = true
        runBlocking { changeSignals.emit(Pair(cache!!, oldValue)) }
        return cache!!
      } finally {
        /// 移除自己作为调用者的身份
        synchronized(StateShared.obsStack) {
          StateShared.obsStack.remove(this)
        }
      }
    }
    return cache!!
  }

  suspend fun set(value: T, force: Boolean = false): Boolean {
    if (setter(value) || force) {
      /// 强制更新值
      get(true)
      /// 向自己的调用者发去通知
      for (ref in refs) {
        ref.beGet(true)
      }
      return true
    }
    return false
  }

  suspend fun updateT(updater: (T?) -> T, force: Boolean = true): Boolean {
    return set(updater(cache), force)
  }

  suspend fun update(updater: (T?) -> Boolean): Boolean {
    if (cache != null) {
      return set(cache!!, updater(cache))
    }
    return false
  }

  suspend fun updateUnit(updater: (T?) -> Unit, force: Boolean = true): Boolean {
    updater(cache)
    if (cache != null) {
      return set(cache!!, force)
    }
    return false
  }

  suspend fun toStream(): Sequence<T> = sequence {
    var waitter: PromiseOut<T>? = null
    val locker = Mutex()
    val cacheList = mutableListOf<T>()
    onChange { data ->
      locker.withLock {
        waitter?.let {
          it.resolve(data.first)
          waitter = null
        }
        if (waitter == null) {
          cacheList.add(data.first)
        }
      }
    }
    try {
      val tmp = runBlocking { get() }
      yield(tmp)
      while (true) {
        while (true) {
          var success: Boolean
          var item: T?
          runBlocking {
            locker.withLock {
              item = cacheList.removeFirst()
              success = (item != null)
            }
          }
          if (!success) {
            break;
          }
          return@sequence yield(item!!)
        }
        runBlocking { locker.lock() }
        val w = PromiseOut<T>()
        waitter = w
        locker.unlock()
        val res = runBlocking { w.waitPromise() }
        return@sequence yield(res)
      }
    } finally {
      locker.unlock()
    }
  }

  override suspend fun beGet(force: Boolean?) {
    get(force)
  }

}